### ISSUE TITLE
[New_Feature]: Count occurences in high cardinality variable.

### ISSUE THREAD

**Issue by coforfe** [ELToulemonde/dataPreparation#6]
Hello Enmanuel-Lin,

I have just seen your very interesting packages announced in CRAN.
Your package offers very interesting possibilities and by using data.table it guarantees high processing speed with low RAM consumption.

Another typical transformation when dealing with datasets is how to reduce the cardinality of some variables (the number of levels they have). With high cardinality it makes almost impossible to use some model techniques.

I have seen that one way to deal with that is to generate another column that represents the number of occurences that each level has in the variable.

This is a reproducible expample:

```r
library(data.table)
library(randNames)

dat_nam <- rand_names(100, gender = "male", nationality = "FR")
head(dat_nam)
```
```r
> head(dat_nam)
# A tibble: 6 x 25
  gender                       email                 dob          registered          phone           cell   nat name.title name.first name.last
   <chr>                       <chr>               <chr>               <chr>          <chr>          <chr> <chr>      <chr>      <chr>     <chr>
1   male    fabien.morel@example.com 1951-12-01 08:23:52 2005-09-24 05:52:23 01-89-87-09-90 06-63-54-70-63    FR         mr     fabien     morel
2   male      loan.henry@example.com 1987-08-31 16:42:02 2008-04-23 18:54:45 03-73-25-83-77 06-90-69-04-55    FR         mr       loan     henry
3   male mathias.garnier@example.com 1983-06-17 03:06:36 2011-11-08 06:17:31 04-24-58-56-01 06-90-89-85-50    FR         mr    mathias   garnier
4   male  valentin.lucas@example.com 1955-01-02 20:43:15 2008-10-31 05:21:31 02-46-30-81-64 06-72-11-22-39    FR         mr   valentin     lucas
5   male      nathan.rey@example.com 1986-10-02 23:24:36 2010-10-13 02:35:53 05-03-40-18-08 06-44-25-63-64    FR         mr     nathan       rey
6   male  matthieu.petit@example.com 1984-10-22 03:10:07 2003-06-02 07:53:58 05-55-19-38-12 06-69-76-48-63    FR         mr   matthieu     petit
# ... with 15 more variables: location.street <chr>, location.city <chr>, location.state <chr>, location.postcode <int>, login.username <chr>,
#   login.password <chr>, login.salt <chr>, login.md5 <chr>, login.sha1 <chr>, login.sha256 <chr>, id.name <chr>, id.value <chr>,
#   picture.large <chr>, picture.medium <chr>, picture.thumbnail <chr>
```
```r
dat_nam <- as.data.table(dat_nam)
dat_red <- dat_nam[, .(name.first, name.last, location.city)]
head(dat_red)
```
```r
  name.first name.last    location.city
1:     fabien     morel  rueil-malmaison
2:       loan     henry          orléans
3:    mathias   garnier      montpellier
4:   valentin     lucas          orléans
5:     nathan       rey aulnay-sous-bois
6:   matthieu     petit           nantes
```

```r
dat_red[, loc_num := .N, by=location.city]
head(dat_red)
```
```r
   name.first name.last    location.city loc_num
1:     fabien     morel  rueil-malmaison       2
2:       loan     henry          orléans       3
3:    mathias   garnier      montpellier       4
4:   valentin     lucas          orléans       3
5:     nathan       rey aulnay-sous-bois       1
6:   matthieu     petit           nantes       4
```

In the last `data.frame`  a new column `loc_num` is added based on the number of occurences of each `location.city`.

I think it could be a nice functionality to include in your package.

Thanks,
Carlos.

**Comment by ELToulemonde** [comment]
Hi,

Thanks for your feed back!

I guess, that this feature should go in a function that would recode a categorical variable in various ways. 
Do you have other way to recode you would like to see implemented? 

Emmanuel-Lin.

**Comment by coforfe** [comment]
Hi,

OK.
Besides counting (absolute number) the number of ocurrences of each string, another possibility would be to calculate the frequencies (percentages).

Thanks,
Carlos.

**Comment by ELToulemonde** [comment]
Hi Carlos,

This feature is now available on CRAN v0.2.

You can do it by performing the following:
```r
library(data.table)
library(randNames)

dat_nam <- rand_names(100, gender = "male", nationality = "FR")

# reduce it 
dat_nam <- as.data.table(dat_nam)
dat_red <- dat_nam[, .(name.first, name.last, location.city)]
head(dat_red)
```
```r
  name.first name.last    location.city
1:     fabien     morel  rueil-malmaison
2:       loan     henry          orléans
3:    mathias   garnier      montpellier
4:   valentin     lucas          orléans
5:     nathan       rey aulnay-sous-bois
6:   matthieu     petit           nantes
```
```r
# Using library dataPreparation
library(dataPreparation)
dat_red <- generateFromCharacter(dat_red, cols = "location.city")

head(dat_red)
```
```r
   name.first name.last   location.city location.city.notnull location.city.num location.city.order
1:       nino fernandez aix-en-provence                 FALSE                 1                   1
2:      timeo    pierre          angers                 FALSE                 3                   2
3:       ryan   clement          angers                 FALSE                 3                   2
4:       malo    muller          angers                 FALSE                 3                   2
5:      erwan     roche      argenteuil                 FALSE                 1                   3
6:        tom   mercier   aubervilliers                 FALSE                 2                   4
```

3 columns have been generated: 

- Was the *location.city* field null?
- Number of occurence of the city (what you asked for)
- The alphabetic order of the city.


In this case, the first feature will be constant but you can get ride of it with *fastFilterVariables*

You can also genrate features from dates and factor using:
- *generateFacotrFromDate*
- *generateDateDiffs*
- *generateFromFactor*

Cheers,

Emmanuel-Lin.
